### DISCUSSION TITLE
DB Update Best Practice

### DISCUSSION THREAD

**Discussion by Ste-denko** [fastapi/fastapi#2561] [Questions] [answered]
When following the FastAPI user guide regarding Databases and SQLAlchemy there are some examples but none for updating data in the DB.
The process I am currently following here is to query the data from the DB (to a SQLAlchemy object-model), transform the data to a Pydantic model/schema, process/change the data within the Pydantic model/schema and then when to put the changed Pydantic-data back into the DB I query again a SQLAlchemy model from the DB, change the needed data and commit the changes.
I cannot reuse the SQLAlchemy model from the beginning since it does not make sense to drag it along throug all the code.
I guess there must be an easier way to update the DB-data.
Since the Pydantic schema is mapped to the database model is there any way to update the database directly from the Pydantic schema?
Or is the simplest way to not transform the data to the Pydantiy schema and just use the SQLAlchemy model?
I would appreciate any best practice info for that?
Thanks in advance.
Cheers!

**Comment by dbanty** [comment, marked as answer]
The way I do it most of the time is something like this:

```python
@app.patch("/items/{item_id}")
def update_item(item_id: int, item: Item, session: Session = Depends(get_session)):
    existing_item = crud.get_item(session=session, item_id=item_id)
    if existing_item is None:
        # Error handling, 404 or whatever
        pass
    crud.update_item(session=session, item=existing_item, updates=item)
```

So I do fetch the sqlalchemy model (existing_item), then update it from the Pydantic model. That `update_item` function would then look something like this:

```python
def update_item(session: Session, item: models.Item, updates: schemas.Item):
    update_data = updates.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(item, key, value)
    session.commit()
```

It's not the prettiest thing ever and you don't get type safety. But it's the best I've come up with for SQLAlchemy.

**Reply by Ste-denko** [reply]
Thank you !

**Reply by ygorpontelo** [reply]
The part about you not getting type safety, pydantic is supposed to check the types and constraints, so why is it not secure?

**Reply by dbanty** [reply]
> The part about you not getting type safety, pydantic is supposed to check the types and constraints, so why is it not secure?

By not being type safe, I mean that the types are completely ignored, so you could set an attribute to an incorrect type and not know unless there's some runtime error in SQLAlchemy.

In the example code that looks like this:

```python
update_data = updates.dict(exclude_unset=True)
for key, value in update_data.items():
    setattr(item, key, value)
```

We know the types of `updates` and `item`, but `update_data` is just a `dict` of type `Dict[str, Any]`. The `setattr` function just takes in `Any, str, Any` for arguments. This means we're setting some arbitrary `Any` value keyed by some unspecific `str` on the object.

So if we had a field on `item` called `price` with type `float` and a field on `updates` called `price` with type `str`, no static checks could help us (in the IDE nor with something like `mypy`) because it would just see us setting some arbitrary attribute to some arbitrary value which is perfectly legal.

On the other hand, if you manually wrote out an update function like this:

```
item.price = updates.price
...
```

Then you'd gain type safety in that static type checkers could tell you you're setting a `float` to a `str` which is probably not what you want.

**Reply by ygorpontelo** [reply]
Oh i see, it's about using static checks. Thanks for the reply!
